llvm/lib/Target/AIE/AIEWawRegRewriter.cpp

```diff
@@ -0,0 +1,386 @@
+        // BlockedPhysRegs. Initially all the used physical registers
+        // from the MBB are blocked, so that replacements do not introduce WAW
+        // dependencies. Additionally, replaced registers are already blocked in
+        // BlockedPhysRegs, so that an additional replacement will not cause a
+        // WAW, which this pass is trying to remove.
+        UsedPhysRegs[VRM->getPhys(Reg)] = true;
+      }
+    }
+  }
+
+  return Modified;
+}
+
+bool AIEWawRegRewriter::isWorthRenaming(
+    const Register &Reg, const BitVector &UsedPhysRegs,
+    const SmallVector<Register, 16> &VRegWithCopies) const {
+  assert(Reg.isVirtual() && MRI->getRegClass(Reg));
+
+  // Only rename registers mapped to a phys reg assigned more than once
+  if (!UsedPhysRegs[VRM->getPhys(Reg)])
+    return false;
+
+  if (!TRI->isVecOrAccRegClass(*(MRI->getRegClass(Reg))))
+    return false;
+
+  return find(VRegWithCopies, Reg) == VRegWithCopies.end();
+}
+
+BitVector
+AIEWawRegRewriter::getDefinedPhysRegs(const MachineBasicBlock *MBB) const {
+  BitVector BlockedPhysRegs(TRI->getNumRegs());
+
+  for (const MachineInstr &MI : *MBB) {
+    for (const MachineOperand &Op : MI.defs()) {
+      MCPhysReg PhysReg = getAssignedPhysReg(Op);
+      if (MCRegister::isPhysicalRegister(PhysReg))
+        addAliasRegs(BlockedPhysRegs, PhysReg);
+    }
+  }
+
+  return BlockedPhysRegs;
+}
+
+bool AIEWawRegRewriter::replaceReg(const Register Reg,
+                                   BitVector &BlockedPhysRegs) {
+  assert(Reg.isVirtual());
+  LLVM_DEBUG(dbgs() << " WAW RegRewriter: Register to replace "
+                    << TRI->getName(VRM->getPhys(Reg)) << "\n");
+
+  MCPhysReg ReplacementPhysReg = getReplacementPhysReg(Reg, BlockedPhysRegs);
+
+  if (!MCRegister::isPhysicalRegister(ReplacementPhysReg))
+    return false;
+
+  LLVM_DEBUG(dbgs() << "     WAW Replacement: Virtual Register "
+                    << Register::virtReg2Index(Reg) << " will replace "
+                    << printReg(VRM->getPhys(Reg), TRI, 0, MRI) << " with "
+                    << printReg(ReplacementPhysReg, TRI, 0, MRI) << '\n');
+
+  const LiveInterval &LI = LIS->getInterval(Reg);
+  LRM->unassign(LI);
+  LRM->assign(LI, ReplacementPhysReg);
+  addAliasRegs(BlockedPhysRegs, ReplacementPhysReg);
+  return true;
+}
+
+MCPhysReg AIEWawRegRewriter::getReplacementPhysReg(
+    const Register Reg, const BitVector &BlockedPhysRegs) const {
+  assert(Reg.isVirtual() && "Reg has to be a virtual register");
+  const TargetRegisterClass *RC = MRI->getRegClass(Reg);
+
+  LiveInterval &LI = LIS->getInterval(Reg);
+  for (const MCPhysReg &PhysReg : RC->getRegisters()) {
+
+    if (BlockedPhysRegs[PhysReg])
+      continue;
+
+    LiveRegMatrix::InterferenceKind IK = LRM->checkInterference(LI, PhysReg);
+    if (IK == LiveRegMatrix::IK_Free)
+      return PhysReg;
+  }
+  return MCRegister::NoRegister;
+}
+
+bool AIEWawRegRewriter::isIdentityCopy(const MachineInstr &MI) const {
+  if (!MI.isCopy())
```

Comment:
Since AIE overloaded this method, you may want to use `TargetInstrInfo::isCopyInstr` here instead of checking for a generic copy instruction.

Reply:
changed to `TargetInstrInfo::isCopyInstr`